packages/tailwindcss/src/plugin-api.ts

```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
         })
       }
     },
+    theme(path) {
+      let cssValue = readFromCss(designSystem.theme, path)
+
+      if (typeof cssValue !== 'object') {
+        return cssValue
+      }
+
+      let result = resolvedConfig.theme ?? {}
+
+      // 1. Convert `path` to an array of parts (segment)
+      // 2. Find the "obeject" that corresponds to the path
+      // 3. Deep merge the object with the resolved theme
+      let parts = segment(path, '.')
+
+      let obj = get(result, parts) ?? {}
+
+      if (typeof obj === 'object') {
+        deepMerge(obj, [cssValue], (a, b) => b)
+        set(result, parts, obj)
+      } else {
+        // This case happens when the CSS theme produces an object for a given
+        // keypath but the v3-style config does not. However we still want the
+        // CSS to "win" in this case so we overwrite the v3-style config value.
+        set(result, parts, cssValue)
+      }
+
+      for (let part of parts) {
+        result = result?.[part]
+      }
+
+      return result
+    },
   }
 }
 
 export function registerPlugins(plugins: Plugin[], designSystem: DesignSystem, ast: AstNode[]) {
-  let pluginApi = buildPluginApi(designSystem, ast)
+  let pluginObjects = []
 
   for (let plugin of plugins) {
     if ('__isOptionsFunction' in plugin) {
       // Happens with `plugin.withOptions()` when no options were passed:
       // e.g. `require("my-plugin")` instead of `require("my-plugin")(options)`
-      plugin().handler(pluginApi)
+      pluginObjects.push(plugin())
     } else if ('handler' in plugin) {
       // Happens with `plugin(…)`:
       // e.g. `require("my-plugin")`
       //
       // or with `plugin.withOptions()` when the user passed options:
       // e.g. `require("my-plugin")(options)`
-      plugin.handler(pluginApi)
+      pluginObjects.push(plugin)
     } else {
       // Just a plain function without using the plugin(…) API
-      plugin(pluginApi)
+      pluginObjects.push({ handler: plugin, config: {} as UserConfig })
+    }
+  }
+
+  // Now merge all the configs and make all that crap work
+  let resolvedConfig = resolveConfig([
+    createCompatabilityConfig(designSystem.theme),
+    ...pluginObjects.map(({ config }) => config ?? {}),
+  ])
```

Comment:
Nothing to add to this PR but I am wondering how we can make this accessible from where we evaluate the CSS `theme()` function. 🤔 Do you have any ideas?

We need the CSS theme function in two places right now:
- In a walk over all property values and `@theme` directives (currently we do our own walk and I think it's fine to do it after plugins are initialized since there are no downstream dependencies right now.
- In another walk on the AST created from utilities

Ideally we don't need to pass this legacy config around in the core of Tailwind. Maybe we can pass the Plugin API around. Another idea is to do the walk after we call register the plugins in this function but that won't allow us to hook it up when we parse the AST from utilities later.

Reply:
We'll handle this in your CSS theme fn PR as that will likely inform the design. I have some ideas on how this will work but need to work through them to see if any of them are good or not.